--- a/configure
+++ b/configure
<<<<<<< HEAD
@@ -241,7 +241,7 @@ int main(int argc, char **argv) {
=======
@@ -259,7 +259,7 @@ int main(int argc, char **argv) {
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 }
 EOF
 
-    $CC -o $TMPDIR/libbpf_test $TMPDIR/libbpf_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS >/dev/null 2>&1
+    $CC -o $TMPDIR/libbpf_test $TMPDIR/libbpf_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS $LDFLAGS >/dev/null 2>&1
     local ret=$?
 
     rm -f $TMPDIR/libbpf_test.c $TMPDIR/libbpf_test
<<<<<<< HEAD
@@ -259,7 +259,7 @@ int main(int argc, char **argv) {
=======
@@ -277,7 +277,7 @@ int main(int argc, char **argv) {
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 }
 EOF
 
-    $CC -o $TMPDIR/libbpf_sec_test $TMPDIR/libbpf_sec_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS >/dev/null 2>&1
+    $CC -o $TMPDIR/libbpf_sec_test $TMPDIR/libbpf_sec_test.c $LIBBPF_CFLAGS $LIBBPF_LDLIBS $LDFLAGS >/dev/null 2>&1
     local ret=$?
 
     rm -f $TMPDIR/libbpf_sec_test.c $TMPDIR/libbpf_sec_test
