<<<<<<< HEAD
--- a/makefile.rules
+++ b/makefile.rules
@@ -101,7 +101,7 @@ OBJ_FILES_KEYGEN =			\
=======
--- elftosb-10.12.01/makefile.rules	2012-03-15 11:01:44.979020178 -0400
+++ elftosb-10.12.01/makefile.rules	2012-03-15 11:01:16.332761989 -0400
@@ -101,7 +101,7 @@
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 	keygen.o
 
 
-LIBS =     -lstdc++
+LIBS =     -lstdc++ -lm
 
 
 ifeq ("${UNAMES}", "Linux")
