--- a/Makefile.am
+++ b/Makefile.am
@@ -1,6 +1,6 @@
 AUTOMAKE_OPTIONS=foreign
 ACLOCAL_AMFLAGS = -I build-aux
-SUBDIRS=doc scripts test
+SUBDIRS=scripts test
 
<<<<<<< HEAD
 noinst_LTLIBRARIES = libcommunicate.la libmacosx.la libfakeroot_time64.la
=======
 noinst_LTLIBRARIES = libcommunicate.la libmacosx.la
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 libcommunicate_la_SOURCES = communicate.c
