--- a/libxtables/xtables.c
+++ b/libxtables/xtables.c
<<<<<<< HEAD
@@ -475,7 +475,7 @@ char *xtables_strdup(const char *s)
 	return dup;
 }
 
-static char *get_modprobe(void)
+__attribute__((unused)) static char *get_modprobe(void)
 {
 	int procfile;
 	char *ret;
@@ -505,6 +505,7 @@ static char *get_modprobe(void)
=======
@@ -403,6 +403,7 @@ static char *get_modprobe(void)
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 
 int xtables_insmod(const char *modname, const char *modprobe, bool quiet)
 {
+#if 0
 	char *buf = NULL;
 	char *argv[4];
 	int status;
<<<<<<< HEAD
@@ -539,6 +540,7 @@ int xtables_insmod(const char *modname,
=======
@@ -437,6 +438,7 @@ int xtables_insmod(const char *modname,
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 	free(buf);
 	if (WIFEXITED(status) && WEXITSTATUS(status) == 0)
 		return 0;
+#endif
 	return -1;
 }
 
