--- a/iptables/xtables-legacy-multi.c
+++ b/iptables/xtables-legacy-multi.c
<<<<<<< HEAD
@@ -28,8 +28,10 @@ static const struct subcommand multi_sub
=======
@@ -32,8 +32,10 @@ static const struct subcommand multi_sub
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
 
 
 #endif
+#ifdef ENABLE_XML
 	{"iptables-xml",        iptables_xml_main},
 	{"xml",                 iptables_xml_main},
+#endif
 #ifdef ENABLE_IPV6
 	{"ip6tables",           ip6tables_main},
 	{"main6",               ip6tables_main},
