--- a/src/mklibs
+++ b/src/mklibs
<<<<<<< HEAD
@@ -269,6 +269,7 @@ def extract_soname(so_file):
=======
@@ -274,6 +274,7 @@ def extract_soname(so_file):
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
     return ""
 
 def multiarch(paths):
+    return paths
     devnull = open('/dev/null', 'w')
     dpkg_architecture = subprocess.Popen(
         ['dpkg-architecture', '-qDEB_HOST_MULTIARCH'],
