--- a/hostapd/Makefile
+++ b/hostapd/Makefile
@@ -315,7 +315,7 @@ NEED_DRAGONFLY=y
 endif
 
 ifdef CONFIG_OWE
-CFLAGS += -DCONFIG_OWE
+CFLAGS += -DCONFIG_OWE -DHOSTAPD_OWE_SUPPORT
 NEED_ECC=y
 NEED_HMAC_SHA256_KDF=y
 NEED_HMAC_SHA384_KDF=y
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -2688,6 +2688,15 @@ static int nl80211_mgmt_subscribe_ap_dev
 					   false) < 0)
 			wpa_printf(MSG_DEBUG,
 				   "nl80211: Failed to subscribe to handle Authentication frames - SAE offload may not work");
+
+#ifdef HOSTAPD_OWE_SUPPORT
+		/* Register for assoc req frames owe mode*/
+		type = (WLAN_FC_TYPE_MGMT << 2) | (WLAN_FC_STYPE_ASSOC_REQ << 4);
+		if (nl80211_register_frame(bss, bss->nl_mgmt, type, NULL, 0,
+					   false) < 0)
+			wpa_printf(MSG_DEBUG,
+				   "nl80211: Failed to subscribe to handle assoc req frames - SAE offload may not work");
+#endif
 	}
 
 	nl80211_mgmt_handle_register_eloop(bss);
