<<<<<<< HEAD
--- a/src/cli-runopts.c
+++ b/src/cli-runopts.c
@@ -340,6 +340,10 @@ void cli_getopts(int argc, char ** argv)
 				case 'z':
 					opts.disable_ip_tos = 1;
 					break;
+				case 'x':
+					/* compatibility with openssh cli
+					 * ("-x" disables X11 forwarding) */
+					break;
 				default:
 					fprintf(stderr,
 						"WARNING: Ignoring unknown option -%c\n", c);
=======
--- a/cli-runopts.c
+++ b/cli-runopts.c
@@ -299,6 +299,8 @@ void cli_getopts(int argc, char ** argv)
 					debug_trace = 1;
 					break;
 #endif
+				case 'x':
+					break;
 				case 'F':
 				case 'e':
 #if !DROPBEAR_USER_ALGO_LIST
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
