<<<<<<< HEAD
--- a/configure.ac
+++ b/configure.ac
@@ -44,11 +44,8 @@ fi
 # LTM_CFLAGS is given to ./configure by the user, 
 # DROPBEAR_LTM_CFLAGS is substituted in the LTM Makefile.in
 DROPBEAR_LTM_CFLAGS="$LTM_CFLAGS"
-if test -z "$DROPBEAR_LTM_CFLAGS"; then
-	DROPBEAR_LTM_CFLAGS="-O3 -funroll-loops -fomit-frame-pointer"
-fi
-AC_MSG_NOTICE(Setting LTM_CFLAGS to $DROPBEAR_LTM_CFLAGS)
-AC_ARG_VAR(LTM_CFLAGS, CFLAGS for bundled libtommath. Default -O3 -funroll-loops -fomit-frame-pointer)
+AC_MSG_NOTICE(Setting LTM_CFLAGS to '$DROPBEAR_LTM_CFLAGS')
+AC_ARG_VAR(LTM_CFLAGS, CFLAGS for bundled libtommath. Defaults to empty string)
 AC_SUBST(DROPBEAR_LTM_CFLAGS)
 
 AC_MSG_NOTICE([Checking if compiler '$CC' supports -Wno-pointer-sign])
--- a/libtomcrypt/src/headers/tomcrypt_dropbear.h
+++ b/libtomcrypt/src/headers/tomcrypt_dropbear.h
@@ -7,8 +7,10 @@
 
 /* Use small code where possible */
 #if DROPBEAR_SMALL_CODE
+#ifndef LTC_SMALL_CODE
 #define LTC_SMALL_CODE
 #endif
+#endif
 
 /* Fewer entries needed */
 #define TAB_SIZE      5
=======
--- a/libtomcrypt/makefile_include.mk
+++ b/libtomcrypt/makefile_include.mk
@@ -94,6 +94,13 @@ endif
 
 LTC_CFLAGS += -Wno-type-limits
 
+ifdef OPENWRT_BUILD
+  ifeq (-Os,$(filter -Os,$(CFLAGS)))
+    LTC_CFLAGS += -DLTC_SMALL_CODE
+  endif
+else
+  ### ! OPENWRT_BUILD
+
 ifdef LTC_DEBUG
 $(info Debug build)
 # compile for DEBUGGING (required for ccmalloc checking!!!)
@@ -121,6 +128,9 @@ endif
 endif # COMPILE_SMALL
 endif # COMPILE_DEBUG
 
+  ### ! OPENWRT_BUILD
+endif
+
 
 ifneq ($(findstring clang,$(CC)),)
 LTC_CFLAGS += -Wno-typedef-redefinition -Wno-tautological-compare -Wno-builtin-requires-header -Wno-missing-field-initializers
--- a/libtommath/makefile_include.mk
+++ b/libtommath/makefile_include.mk
@@ -70,6 +70,9 @@ else
 LTM_CFLAGS += -Wsystem-headers
 endif
 
+ifndef OPENWRT_BUILD
+  ### ! OPENWRT_BUILD
+
 ifdef COMPILE_DEBUG
 #debug
 LTM_CFLAGS += -g3
@@ -90,6 +93,9 @@ endif
 
 endif # COMPILE_SIZE
 
+  ### ! OPENWRT_BUILD
+endif
+
 ifneq ($(findstring clang,$(CC)),)
 LTM_CFLAGS += -Wno-typedef-redefinition -Wno-tautological-compare -Wno-builtin-requires-header
 endif
>>>>>>> 712839d4c6 (Removed unwanted submodules from index)
